crates/modules/process-monitor/probes.bpf.c

```diff
@@ -193,36 +195,68 @@ static __always_inline int get_container_info(struct task_struct *cur_tsk, char
     return FAILED_READ_MEMORY_CGROUP_ID;
   }
 
-  const char *name = BPF_CORE_READ(cur_tsk, cgroups, subsys[cgrp_id], cgroup, kn, name);
-  if (bpf_probe_read_kernel_str(buf, sz, name) < 0)
+  struct kernfs_node *kn = BPF_CORE_READ(cur_tsk, cgroups, subsys[cgrp_id], cgroup, kn);
+  const char *name = BPF_CORE_READ(kn, name);
+  if (bpf_probe_read_kernel_str(buf, CONTAINER_ID_MAX_BUF, name) < 0)
   {
     LOG_ERROR("failed to get kernfs node name: %s\n", buf);
     return FAILED_READ_CGROUP_NAME;
   }
 
-  // Docker case
+  struct kernfs_node *parent_kn = BPF_CORE_READ(kn, parent);
+  const char *parent_name = NULL;
+  if (parent_kn != NULL)
+  {
+    parent_name = BPF_CORE_READ(kn, parent, name);
+    if (bpf_probe_read_kernel_str(buf_parent, CONTAINER_ID_MAX_BUF,
+                                  parent_name) < 0)
+    {
+      LOG_ERROR("failed to get parent kernfs node name: %s\n", buf_parent);
+      return FAILED_READ_PARENT_CGROUP_NAME;
+    }
+  }
+
+  // For Docker containers, there are two cgroup path formats:
+  //
+  // 1. `0::/system.slice/docker-d4ea646fc22c701dbb146e52db4e9125dcca2eebb2f5552f90fedbb28a0f0716.scope`
+  //    This is typical for systemd cgroup driver. In this case, the last
+  //    node contains the `docker-` prefix followed by the container ID.
+  // 2. `0::/docker/2cf2e0be458a80acb354c953f7bb03de5d2d277dfcb8ebaa6575f95668a0c15f`
+  //    This is typical for cgroupfs driver. In this case, the parent node is
+  //    `docker` and the child node contains the container ID.
+
+  // Case 1.
   if (STRNCMP(buf, 7, "docker-") == 0)
   {
     *offset = 7;
     return DOCKER_CONTAINER_ENGINE;
   }
 
+  // Case 2.
+  //
+  // The check for NULL character is needed to make sure it is the full kernfs
+  // node name.
+  if (STRNCMP(buf_parent, 6, "docker") == 0 && buf_parent[6] == '\0')
+  {
+    // The last node is unprefixed, it contains just container ID.
+    *offset = 0;
+    return DOCKER_CONTAINER_ENGINE;
+  }
+
   // Podman case
   //
   // the check for NULL character is needed to avoid collisions with
   // `containerd-` prefixed cgroup name
   if (STRNCMP(buf, 9, "container") == 0 && buf[9] == '\0')
   {
-
-    const char *parent_name = BPF_CORE_READ(cur_tsk, cgroups, subsys[cgrp_id], cgroup, kn, parent, name);
-
-    if (bpf_probe_read_kernel_str(buf, sz, parent_name) < 0)
+    // Read `parent_name` to the main buffer `buf`.
+    if (parent_name == NULL || bpf_probe_read_kernel_str(buf, CONTAINER_ID_MAX_BUF, parent_name) < 0)
     {
       LOG_ERROR("failed to get parent kernfs node name: %s\n", buf);
       return FAILED_READ_PARENT_CGROUP_NAME;
     }
 
-    if (STRNCMP(buf, 7, "libpod-") == 0)
+    if (STRNCMP(buf_parent, 7, "libpod-") == 0)
```

Comment:
here you should use `buf` because at line 253 you read into `buf` and the output should be there.

this error makes me thinking if `char buf_parent[CONTAINER_ID_MAX_BUF];` should be internal to this function. the only meaningful information will always be in `buf`